rpc: flake in TestConnHealth

[Here](https://teamcity.cockroachdb.com/viewLog.html?buildId=7478184&buildTypeId=Cockroach_BazelExtendedCi)

```
Failed
=== RUN   TestConnHealth
I221112 19:09:03.598670 32 2@rpc/context.go:2158  [rnode=1,raddr=127.0.0.1:42409,class=default,rpc] 11  connection is now ready
I221112 19:09:03.700270 32 2@rpc/context.go:1998  [rnode=1,raddr=127.0.0.1:42409,class=default,rpc] 12  closing connection after: grpc: boom [code 2/Unknown]
I221112 19:09:03.700318 148 1@circuitbreaker/circuitbreaker.go:322  [-] 13  circuitbreaker: rpc  [n1] tripped: failed to connect to n1 at 127.0.0.1:42409: grpc: boom [code 2/Unknown]
I221112 19:09:03.700401 148 1@circuitbreaker/circuitbreaker.go:447  [-] 14  circuitbreaker: rpc  [n1] event: BreakerTripped
W221112 19:09:03.700415 148 2@rpc/nodedialer/nodedialer.go:191  [-] 15  unable to connect to n1: failed to connect to n1 at 127.0.0.1:42409: grpc: boom [code 2/Unknown]
    nodedialer_test.go:160: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/9297/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/rpc/nodedialer/nodedialer_test_/nodedialer_test.runfiles/com_github_cockroachdb_cockroach/pkg/rpc/nodedialer/nodedialer_test.go:160
        	Error:      	Received unexpected error:
        	            	failed to connect to n1 at 127.0.0.1:42409: grpc: boom [code 2/Unknown]
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).dial
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:200
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).DialNoBreaker
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:137
        	            	  | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.TestConnHealth
        	            	  | 	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer_test.go:159
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1446
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_arm64.s:1172
        	            	Wraps: (2) failed to connect to n1 at 127.0.0.1:42409
        	            	Wraps: (3) grpc: boom [code 2/Unknown]
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *status.Error
        	Test:       	TestConnHealth
--- FAIL: TestConnHealth (0.11s)
```


Jira issue: CRDB-21434

--- a/pkg/rpc/nodedialer/nodedialer_test.go
+++ b/pkg/rpc/nodedialer/nodedialer_test.go
@@ -156,8 +156,16 @@ func TestConnHealth(t *testing.T) {
 	// dials the node.
 	hb.setErr(nil)
 	{
-		_, err := nd.DialNoBreaker(ctx, staticNodeID, rpc.DefaultClass)
-		require.NoError(t, err)
+		// In practice this dial almost always immediately succeeds. However,
+		// because the previous connection was marked as unhealthy and might
+		// still just be about to be removed, the DialNoBreaker might pull
+		// the broken connection out of the pool and fail on it.
+		//
+		// See: https://github.com/cockroachdb/cockroach/issues/91798
+		require.Eventually(t, func() bool {
+			_, err := nd.DialNoBreaker(ctx, staticNodeID, rpc.DefaultClass)
+			return err == nil
+		}, 10*time.Second, time.Millisecond)
 		require.NoError(t, nd.ConnHealth(staticNodeID, rpc.DefaultClass))
 	}
 

